File: aspnet/src/ShowAd.AspNetCore/AccessPolicy/CrawlerPolicy.cs
using Microsoft.AspNetCore.Http;

namespace ShowAd.AspNetCore.AccessPolicy;

public delegate bool ReverseDnsVerifier(string ip, string family);

public class CrawlerPolicy
{
    public bool Enabled { get; set; }

    /// <summary>Restrict to a subset of families. Null = all known families.</summary>
    public IList<string>? Families { get; set; }

    /// <summary>Override default UA fragments per family (lowercase substrings).</summary>
    public IDictionary<string, IList<string>>? UserAgents { get; set; }

    /// <summary>CIDR ranges per family for IP verification.</summary>
    public IDictionary<string, IList<string>> FamilyCidrs { get; set; } = new Dictionary<string, IList<string>>(StringComparer.OrdinalIgnoreCase);

    /// <summary>Trust the Cloudflare verified-bot signal (CF-Verified-Bot header).</summary>
    public bool AllowCloudflareVerifiedBot { get; set; }

    /// <summary>Optional callback for reverse-DNS verification.</summary>
    public ReverseDnsVerifier? ReverseDnsVerifier { get; set; }

    /// <summary>Default crawler UA fragments per family. UA alone never bypasses.</summary>
    public static readonly IReadOnlyDictionary<string, IReadOnlyList<string>> DefaultUserAgents =
        new Dictionary<string, IReadOnlyList<string>>(StringComparer.OrdinalIgnoreCase)
        {
            ["google"] = new[] { "googlebot", "google-inspectiontool", "apis-google" },
            ["bing"] = new[] { "bingbot" },
            ["duckduckgo"] = new[] { "duckduckbot" },
            ["yandex"] = new[] { "yandexbot" },
            ["baidu"] = new[] { "baiduspider" },
            ["openai"] = new[] { "gptbot", "chatgpt-user", "oai-searchbot" },
            ["anthropic"] = new[] { "claudebot", "anthropic-ai" },
            ["perplexity"] = new[] { "perplexitybot" },
            ["commoncrawl"] = new[] { "ccbot" },
            ["facebook"] = new[] { "facebookexternalhit", "facebot" },
            ["twitter"] = new[] { "twitterbot" },
            ["linkedin"] = new[] { "linkedinbot" },
        };
}

public class AccessPolicyOptions
{
    /// <summary>Trusted edge headers carrying the real client IP.</summary>
    public IList<string> TrustedIpHeaders { get; set; } = new List<string>();

    /// <summary>CIDR allowlist evaluated against the resolved client IP.</summary>
    public IList<string> AllowCidrs { get; set; } = new List<string>();

    /// <summary>Verified-crawler policy. UA alone never bypasses.</summary>
    public CrawlerPolicy? Crawler { get; set; }

    /// <summary>Publisher callback (premium users, app sessions, ...).</summary>
    public Func<HttpContext, AccessPolicyContext, AccessPolicyDecision>? BeforeProtect { get; set; }
}

public sealed class AccessPolicyContext
{
    public string? ClientIp { get; init; }
    public string UserAgent { get; init; } = string.Empty;
}

public enum AccessPolicyAction { Continue, Allow, Redirect }

public sealed class AccessPolicyDecision
{
    public AccessPolicyAction Action { get; init; } = AccessPolicyAction.Continue;
    public string? Reason { get; init; }
    public string? RedirectUrl { get; init; }

    public static AccessPolicyDecision Continue(string? reason = null) =>
        new() { Action = AccessPolicyAction.Continue, Reason = reason };

    public static AccessPolicyDecision Allow(string? reason = null) =>
        new() { Action = AccessPolicyAction.Allow, Reason = reason };

    public static AccessPolicyDecision Redirect(string? url = null, string? reason = null) =>
        new() { Action = AccessPolicyAction.Redirect, Reason = reason, RedirectUrl = url };
}

File: springboot/src/main/java/io/proofmark/showad/access/CrawlerPolicy.java
package io.proofmark.showad.access;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/**
 * Verified crawler configuration. UA matches alone never bypass verification:
 * an IP must additionally match either the configured CIDR list for the family
 * or the Cloudflare verified-bot header (when enabled).
 */
public final class CrawlerPolicy {

    public static final Map<String, List<String>> DEFAULT_USER_AGENTS;

    static {
        Map<String, List<String>> defaults = new LinkedHashMap<>();
        defaults.put("google", List.of("googlebot", "google-inspectiontool", "apis-google"));
        defaults.put("bing", List.of("bingbot"));
        defaults.put("duckduckgo", List.of("duckduckbot"));
        defaults.put("yandex", List.of("yandexbot"));
        defaults.put("baidu", List.of("baiduspider"));
        defaults.put("openai", List.of("gptbot", "chatgpt-user", "oai-searchbot"));
        defaults.put("anthropic", List.of("claudebot", "anthropic-ai"));
        defaults.put("perplexity", List.of("perplexitybot"));
        defaults.put("commoncrawl", List.of("ccbot"));
        defaults.put("facebook", List.of("facebookexternalhit", "facebot"));
        defaults.put("twitter", List.of("twitterbot"));
        defaults.put("linkedin", List.of("linkedinbot"));
        DEFAULT_USER_AGENTS = Collections.unmodifiableMap(defaults);
    }

    private final boolean enabled;
    private final boolean allowCloudflareVerifiedBot;
    private final List<String> families;
    private final Map<String, List<String>> userAgents;
    private final Map<String, List<String>> familyCidrs;

    public CrawlerPolicy(
        boolean enabled,
        boolean allowCloudflareVerifiedBot,
        List<String> families,
        Map<String, List<String>> userAgents,
        Map<String, List<String>> familyCidrs
    ) {
        this.enabled = enabled;
        this.allowCloudflareVerifiedBot = allowCloudflareVerifiedBot;
        this.userAgents = (userAgents == null || userAgents.isEmpty())
            ? DEFAULT_USER_AGENTS
            : Map.copyOf(userAgents);
        this.families = (families == null || families.isEmpty())
            ? List.copyOf(this.userAgents.keySet())
            : List.copyOf(families);
        this.familyCidrs = familyCidrs == null ? Map.of() : Map.copyOf(familyCidrs);
    }

    public boolean isEnabled() { return enabled; }
    public boolean isAllowCloudflareVerifiedBot() { return allowCloudflareVerifiedBot; }
    public List<String> getFamilies() { return families; }
    public Map<String, List<String>> getUserAgents() { return userAgents; }
    public Map<String, List<String>> getFamilyCidrs() { return familyCidrs; }

    public String matchFamily(String userAgent) {
        if (userAgent == null || userAgent.isBlank()) {
            return null;
        }
        String needle = userAgent.toLowerCase(Locale.ROOT);
        for (String family : families) {
            List<String> fragments = userAgents.getOrDefault(family, List.of());
            for (String fragment : fragments) {
                if (fragment != null && !fragment.isEmpty()
                    && needle.contains(fragment.toLowerCase(Locale.ROOT))) {
                    return family;
                }
            }
        }
        return null;
    }
}
